### PR TITLE
add evaluation under adversarial patch attack

### PR DESCRIPTION
**Pull request by JindongGu**
# Task Description
Add evaluation under adversarial patch attack

## Dataset Creation
Standard ImageNet Validation Dataset

## Evaluation Metrics
Standard Accuracy

## Expected Insights/Relevance
The robustness to patch attack

# Access
No available

## Data
No

## License
No available

### CODE REVIEW

File: examples/imagenet_adv_patch.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+from shifthappens.tasks.task_result import TaskResult
+
+
+@sh_benchmark.register_task(
+    name="ImageNet_Adv_Patch", relative_data_folder="imagenet_val/", standalone=True
+)
+
+
+@dataclasses.dataclass
+class ImageNet_Adv_Patch(Task):
```
Comment (stes) [inline examples/imagenet_adv_patch.py:36]:
```suggestion
@sh_benchmark.register_task(
    name="ImageNet_Adv_Patch", relative_data_folder="imagenet_val/", standalone=True
)
@dataclasses.dataclass
class ImageNet_Adv_Patch(Task):
```

File: examples/imagenet_adv_patch.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import os
+
+import numpy as np
+import torch
+import torchvision.datasets as tv_datasets
+import torchvision.transforms as tv_transforms
+import torch.nn as nn
+
+import sys
+sys.path.append('./')
```
Comment (stes) [inline examples/imagenet_adv_patch.py:13]:
```suggestion
```

This should be possible to be removed once the changes required via the contribution guide (i.e. placing the code into its own subpackage) are addressed.

File: examples/imagenet_adv_patch.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        return TaskResult(
+            accuracy=accuracy, summary_metrics={Metric.Robustness: "accuracy"}
+        )
+
+
+if __name__ == "__main__":
+    from shifthappens.models.torchvision import ResNet18
+
+    sh_benchmark.evaluate_model(ResNet18(device="cuda:0", max_batch_size=8), "../")
```
Comment (stes) [inline examples/imagenet_adv_patch.py:161]: This can be removed.
```suggestion
```

**Review by stes — CHANGES_REQUESTED** [review]
Hi @JindongGu , thanks a lot for your contribution. We are working towards the first release of the benchmark and for this, released our [contribution guide](https://github.com/shift-happens-benchmark/icml-2022/blob/main/CONTRIBUTING.rst) with the aim of unifying the implementation of all contributions.

Please have a look, the required changes should be relatively small and are mostly concerns with placing the files within the package, and adding documentation.

From reviewing your PR, the main required changes are:

- [ ] Place the files into their own package inside the `tasks/` folder (@georgepachitariu )
- [ ] Add a README and optionally citation info
- [ ] We are also looking into the best ways to perform unit or integration tests and would be happy to get some feedback on how this would be most effective for you benchmark. Would it be possible to add a "dry-run/demo" option to the benchmark, and e.g. run it only on a small subset of the data? No immediate action/update is required right now, we are first gathering opinions from all authors on this.

Thanks a lot for your contribution, and please ping me if you have any questions. If you have technical questions, also feel free to ping @kotekjedi who is happy to help with implementation details.

**Comment by stes** [comment]
Hi!

This is an automated report for the documentation coverage of your PR. To provide good user experience for the benchmark, we aim to cover most functions with doc strings, using the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

We would greatly appreciate your help in assuring that docstring coverage is sufficiently high for your contribution.

The current coverage is:
| Name                         |     Total |     Miss |     Cover |     Cover% |
|------------------------------|-----------|----------|-----------|------------|
| imagenet_adv_patch.py        |         3 |        2 |         1 |        33% |
|------------------------------|-----------|----------|-----------|------------|
| TOTAL                        |         3 |        2 |         1 |      33.3% |
RESULT: FAILED (minimum: 80.0%, actual: 33.3%)

You can reproduce this result by running

```
interrogate -v --ignore-nested-functions --ignore-private --ignore-semiprivate --ignore-init-method
```

on the files you added to your PR.

File: examples/imagenet_adv_patch.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+import torchvision.transforms as tv_transforms
+import torch.nn as nn
+
+import sys
+sys.path.append('./')
+
+import shifthappens.models.base as sh_models
+import shifthappens.data.base as sh_data
+import shifthappens.data.torch as sh_data_torch
+import shifthappens.utils as sh_utils
```
Comment (kotekjedi) [inline examples/imagenet_adv_patch.py:18]:
```suggestion
import shifthappens.models.base as sh_models
import shifthappens.data.base as sh_data
import shifthappens.data.torch as sh_data_torch
import shifthappens.utils as sh_utils
```
Should we remove this? These imports are never used.

File: examples/imagenet_adv_patch.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+from shifthappens.models import base as sh_models
+from shifthappens.models.base import PredictionTargets
+from shifthappens.tasks.base import Task
+from shifthappens.tasks.base import parameter
+from shifthappens.tasks.metrics import Metric
+from shifthappens.tasks.task_result import TaskResult
+
+
+@sh_benchmark.register_task(
+    name="ImageNet_Adv_Patch", relative_data_folder="imagenet_val/", standalone=True
```
Comment (kotekjedi) [inline examples/imagenet_adv_patch.py:31]:
```suggestion
    name="ImageNet_Adv_Patch", relative_data_folder="imagenet_adv_patch", standalone=True
```
We shouldn't use a path name here which is dependent on imagenet validation path.

File: examples/imagenet_adv_patch.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+    patch_size: int = parameter(
+        default=16,
+        options=(16, 32, 64),
+        description="the number of attack iterations",
+    )
+
+
+    def setup(self):
+        dataset_folder = os.path.join(self.data_root, "./")
```
Comment (kotekjedi) [inline examples/imagenet_adv_patch.py:64]: Please, update your PR with a current version of the main.
Here you could use `dataset_folder = shifthappens.config.imagenet_validation_path`.

Or even replace whole data loader creation part with `shifthappens.data.inagenet.get_imagenet_validation_loader`.

File: examples/imagenet_adv_patch.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        if not os.path.exists(dataset_folder):
+            print('The specified data path does not exist. Please Specify the data path correctly !!!')
+            exit()
```
Comment (kotekjedi) [inline examples/imagenet_adv_patch.py:67]:
```suggestion
```

File: examples/imagenet_adv_patch.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+        all_predicted_labels_list = []
+        for batch in dataloader:
+            processed_batch = self._get_adv_patch_images(model, batch, device)
+            predictions = self._eval_preprocessed_batch(model, processed_batch, device)
+            all_predicted_labels_list.append(predictions.class_labels)
+
+        all_predicted_labels = np.concatenate(all_predicted_labels_list, 0)
+
+        accuracy = all_predicted_labels == np.array(self.ch_dataset.targets)
```
Comment (kotekjedi) [inline examples/imagenet_adv_patch.py:151]: Here accuracy is an array of True and False values: should we do `accuracy = np.mean(all_predicted_labels == np.array(self.ch_dataset.targets))`
